insert_string_tpl.h

```diff
@@ -82,20 +82,16 @@ ZLIB_INTERNAL Pos INSERT_STRING(deflate_state *const s, const uint32_t str, uint
         val |= ((uint32_t)(strstart[2]) << 16);
         val |= ((uint32_t)(strstart[3]) << 24);
 #endif
-
         UPDATE_HASH(s, h, val);
         hm = h & s->hash_mask;
 
-        Pos head = s->head[hm];
-        if (head != idx) {
+        head = s->head[hm];
+        if (LIKELY(head != idx)) {
             s->prev[idx & s->w_mask] = head;
             s->head[hm] = idx;
-            if (strstart == strend)
-                ret = head;
-        } else if (strstart == strend) {
-            ret = idx;
         }
     }
-    return ret;
+
+    return head;
```

Comment:
```head``` should not be returned if there was no updating of the hash tables as it can be garbage in case of recent window slide.

Reply:
1. If count > 1 (it always is), the for loop will set head.
2. strstart is always == strend after the for loop has completed, since we never exit it until they are equal.
3. If head != idx, we return head, if head == idx, we return idx (might as well return head since they are the same)

The only corner-case I see is with count==0, but that that is already handled specially above.

I don't see how this can currently return 0, can you give an example of that?

Reply:
Sorry, fixed typo, had = instead of >= in point #1 of my previous comment..

Reply:
In the for loop, s->head[hm] can be garbage as when the window is slided, the hash table is not cleared, so anything that is still in window is no longer valid. It would be really slow to clear whole hash table and recalculate all chains, so on first call to insert_string() the old value was discarded. On subsequent calls there is no more garbage.

If we don't ignore the initial contents after window slide, we can get future matches in for example deflate_medium and that will cause assertion failure in debug build.

Reply:
But how does that actually happen in this function? Please walk me through it, because I don't see how.
Has this been lost and is already a bug?

Reply:
As far as I understand it, after window is slided, the next call to insert_string is from "next match", not "current match" which should be at the start of the window... When it tries to compare the matches, it gets zero offset, because the previous call to insert_match before the slide was also from the "next match". As such the current and previous match become same offset on next check of same hash chain.

If we set the previous offset as 0 (return value of insert_string) when the new and old values match inside insert_string, we actually truncate the hash chain as 0 means end of chain. This doesn't cause assertion failure as the "previous" value in the hash chain is now correct as it was not replaced (the previous value was slided in fill_window, but is guaranteed to be valid).

Reply:
I understand how this would be beneficial, but I cannot see how the current code will ever return `0`. As far as I can see, it always returns either `head` or `idx` (same as head), or with count=0 it returns `s->prev[str & s->w_mask];`.

So my question is how do you trace through this function in a way that `ret` ends up being `0`? IE: How does this function detect that the windows has been slided and a `0` should be returned?

And if it currently cannot return `0` (as it looks like to me), is this actually a hidden bug?

Reply:
I haven't followed all the pull requests that you have merged in, so there might be a previous regression that got unnoticed.

Reply:
@mtl1979 I am not following your explanation very well. I thought we don't use hash-chains aka rolling hash because we are only ever hashing 4 bytes at a time.

Reply:
I'm not talking about linear hashing chain, I'm talking about how we store indexes that have same hash value. That is essential when we want to know if two indexes have same hash value or not... If we store same index for both the "latest" and "previous" hash tables, then we are basically wasting the "previous" hash table.

When the hash tables are initialized, "previous" hash table should be all zeroes, until the "latest" hash table is populated, when the "old" value of "latest" will be moved to "previous" hash table.... as on first iteration (before first window slide) the old value is 0, there is no problem.

Reply:
That is a great explanation but I'm still failing to see how it applies to this code.

Last time I looked, the return value from insert_string is not even used by anything.

Reply:
So we should change the function signature so we don't return anything?

Reply:
Looks like we could do that, since the return value is only ever used when calling quick_insert_string.
Doing some tests now to see whether it has any effect on benchmarks, it does reduce compiled text size by 64, so it should help a bit.

Reply:
I think we have split fixing the function to enough separate pull requests, so we should do one final cleanup on this function and then we don't touch it before first release...

Reply:
The nice thing now is quick_insert_string and insert_string are very close to identical and much cleaner.